File: mobile-engage/src/main/java/com/emarsys/mobileengage/responsehandler/MobileEngageClientStateResponseHandler.java
package com.emarsys.mobileengage.responsehandler;

import com.emarsys.core.response.AbstractResponseHandler;
import com.emarsys.core.response.ResponseModel;
import com.emarsys.core.storage.Storage;
import com.emarsys.core.util.Assert;
import com.emarsys.mobileengage.util.RequestUrlUtils;

public class MobileEngageClientStateResponseHandler extends AbstractResponseHandler {

    private static final String X_CLIENT_STATE = "X-CLIENT-STATE";

    private final Storage<String> clientStateStorage;

    public MobileEngageClientStateResponseHandler(Storage<String> clientStateStorage) {
        Assert.notNull(clientStateStorage, "ClientStateStorage must not be null!");

        this.clientStateStorage = clientStateStorage;
    }

    @Override
    protected boolean shouldHandleResponse(ResponseModel responseModel) {
        boolean isMobileEngageRequest = RequestUrlUtils.isMobileEngageRequest(responseModel.getRequestModel());
        boolean hasClientState = getClientState(responseModel) != null;

        return isMobileEngageRequest && hasClientState;
    }

    @Override
    protected void handleResponse(ResponseModel responseModel) {
        clientStateStorage.set(getClientState(responseModel));
    }

    private String getClientState(ResponseModel responseModel) {
        return responseModel.getHeaders().get(X_CLIENT_STATE);
    }
}

File: mobile-engage/src/main/java/com/emarsys/mobileengage/responsehandler/MobileEngageClientStateResponseHandler.kt
package com.emarsys.mobileengage.responsehandler

import com.emarsys.core.Mockable
import com.emarsys.core.response.AbstractResponseHandler
import com.emarsys.core.response.ResponseModel
import com.emarsys.core.storage.Storage
import com.emarsys.core.util.getCaseInsensitive
import com.emarsys.mobileengage.util.RequestModelHelper

@Mockable
class MobileEngageClientStateResponseHandler(private val clientStateStorage: Storage<String?>,
                                             private val requestModelHelper: RequestModelHelper) : AbstractResponseHandler() {

    companion object {
        private const val X_CLIENT_STATE = "X-Client-State"
    }

    override fun shouldHandleResponse(responseModel: ResponseModel): Boolean {
        val hasClientState = getClientState(responseModel) != null

        return requestModelHelper.isMobileEngageRequest(responseModel.requestModel) && hasClientState
    }

    override fun handleResponse(responseModel: ResponseModel) {
        clientStateStorage.set(getClientState(responseModel))
    }

    private fun getClientState(responseModel: ResponseModel): String? {
        return responseModel.headers.getCaseInsensitive(X_CLIENT_STATE)
    }
}
